src/common/bip352.cpp

```diff
@@ -0,0 +1,384 @@
+// Copyright (c) 2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <addresstype.h>
+#include <common/bip352.h>
+#include <optional>
+#include <primitives/transaction.h>
+#include <secp256k1_extrakeys.h>
+#include <pubkey.h>
+#include <secp256k1.h>
+#include <span.h>
+
+#include <secp256k1_silentpayments.h>
+#include <streams.h>
+#include <uint256.h>
+#include <logging.h>
+#include <policy/policy.h>
+#include <script/sign.h>
+#include <script/solver.h>
+#include <script/script_error.h>
+#include <util/check.h>
+#include <util/strencodings.h>
+
+extern secp256k1_context* secp256k1_context_sign; // TODO: this is hacky, is there a better solution?
+
+namespace bip352 {
+
+class PublicDataImpl
+{
+private:
+    //! The actual secnonce itself
+    std::unique_ptr<secp256k1_silentpayments_recipient_public_data> m_public_data;
+
+public:
+    PublicDataImpl() : m_public_data{std::make_unique<secp256k1_silentpayments_recipient_public_data>()} {}
+
+    // Delete copy constructors
+    PublicDataImpl(const PublicDataImpl&) = delete;
+    PublicDataImpl& operator=(const PublicDataImpl&) = delete;
+
+    secp256k1_silentpayments_recipient_public_data* Get() const { return m_public_data.get(); }
+};
+
+PublicData::PublicData() : m_impl{std::make_unique<PublicDataImpl>()} {}
+
+PublicData::PublicData(PublicData&&) noexcept = default;
+PublicData& PublicData::operator=(PublicData&&) noexcept = default;
+
+PublicData::~PublicData() = default;
+
+secp256k1_silentpayments_recipient_public_data* PublicData::Get() const
+{
+    return m_impl->Get();
+}
+
+std::optional<PubKey> GetPubKeyFromInput(const CTxIn& txin, const CScript& spk)
+{
+    std::vector<std::vector<unsigned char>> solutions;
+    TxoutType type = Solver(spk, solutions);
+    PubKey pubkey;
+    if (type == TxoutType::WITNESS_V1_TAPROOT) {
+        // Check for H point in script path spend
+        if (txin.scriptWitness.stack.size() > 1) {
+            // Check for annex
+            bool has_annex = txin.scriptWitness.stack.back()[0] == ANNEX_TAG;
+            size_t post_annex_size = txin.scriptWitness.stack.size() - (has_annex ? 1 : 0);
+            if (post_annex_size > 1) {
+                // Actually a script path spend
+                const std::vector<unsigned char>& control = txin.scriptWitness.stack.at(post_annex_size - 1);
+                Assert(control.size() >= 33);
+                if (std::equal(WitnessV1Taproot::NUMS_H.begin(), WitnessV1Taproot::NUMS_H.end(), control.begin() + 1)) {
+                    // Skip script path with H internal key
+                    return std::nullopt;
+                }
+            }
+        }
+
+        pubkey = XOnlyPubKey{solutions[0]};
+    } else if (type == TxoutType::WITNESS_V0_KEYHASH) {
+        pubkey = CPubKey{txin.scriptWitness.stack.back()};
```

Comment:
https://github.com/bitcoin/bitcoin/pull/28122/commits/87a6ea213301342afbdd37876939c1eeb142b8e3: This will cause a segfault when `scriptWitness` is empty. `scriptWitness` can be empty if the tranasaction hasn't been signed yet. @macgyver reached this edge case while testing #32966 see https://github.com/bitcoin/bitcoin/pull/32966#discussion_r2277212074